File: leetcode/301. Remove Invalid Parentheses.py
from collections import deque
class Solution:
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        def isValid(element):
            cnt = 0
            for c in element:
                if c== '(':
                    cnt += 1
                elif c == ')':
                    cnt -= 1
                    if cnt < 0:
                        return False
            return cnt == 0
        
        q = deque([s])
        res = []
        visited = set()
        maxlen = 0
        
        while len(q) != 0:
            element = q.popleft()
            if len(element) < maxlen:
                break
            if isValid(element):
                res.append(element)
                maxlen = len(element)
                continue
            for i,c in enumerate(element):
                if c != '(' and c != ')':
                    continue
                sub = element[:i] + element[i+1:]
                if sub in visited:
                    continue
                visited.add(sub)
                q.append(sub)
        return res
                

File: leetcode/301. Remove Invalid Parentheses.kt
class Solution {
    fun removeInvalidParentheses(s: String): List<String> {
        fun isValid(s: String): Boolean {
            var left = 0
            for (c in s) {
                when (c) {
                    '(' -> left++
                    ')' -> left--
                }
                if (left < 0) return false
            }
            return left == 0
        }
        val dq = ArrayDeque<String>(listOf(s))
        val visited = mutableSetOf<String>(s)
        val res = mutableListOf<String>()
        while (dq.size != 0) {
            val s = dq.removeFirst()
            if (res.size != 0 && s.length < res[0].length) break
            if (isValid(s)) {
                res.add(s)
                continue
            }
            for (i in 0..s.length-1) {
                val sub = s.substring(0, i) + s.substring(i+1, s.length)
                if (sub in visited) continue
                visited.add(sub)
                dq.addLast(sub)
            }
        }
        return res
    }
}
